File: lib/cbor.ts
// This file implements a tiny subset of CTAP2's subset of CBOR, in order to
// encode and decode WebAuthn identities.
//
// Only major types 0 (unsigned integer), 2 (byte strings), 3 (text strings),
// and 4 (arrays, only containing text strings) are supported. Arguments are
// limited to 16-bit values.
//
// See https://www.imperialviolet.org/tourofwebauthn/tourofwebauthn.html#cbor.

function readTypeAndArgument(b: Uint8Array): [number, number, Uint8Array] {
    if (b.length === 0) {
        throw Error("cbor: unexpected EOF")
    }
    const major = b[0] >> 5
    const minor = b[0] & 0x1f
    if (minor <= 23) {
        return [major, minor, b.subarray(1)]
    }
    if (minor === 24) {
        if (b.length < 2) {
            throw Error("cbor: unexpected EOF")
        }
        return [major, b[1], b.subarray(2)]
    }
    if (minor === 25) {
        if (b.length < 3) {
            throw Error("cbor: unexpected EOF")
        }
        return [major, (b[1] << 8) | b[2], b.subarray(3)]
    }
    throw Error("cbor: unsupported argument encoding")
}

export function readUint(b: Uint8Array): [number, Uint8Array] {
    const [major, minor, rest] = readTypeAndArgument(b)
    if (major !== 0) {
        throw Error("cbor: expected unsigned integer")
    }
    return [minor, rest]
}

export function readByteString(b: Uint8Array): [Uint8Array, Uint8Array] {
    const [major, minor, rest] = readTypeAndArgument(b)
    if (major !== 2) {
        throw Error("cbor: expected byte string")
    }
    if (minor > rest.length) {
        throw Error("cbor: unexpected EOF")
    }
    return [rest.subarray(0, minor), rest.subarray(minor)]
}

export function readTextString(b: Uint8Array): [string, Uint8Array] {
    const [major, minor, rest] = readTypeAndArgument(b)
    if (major !== 3) {
        throw Error("cbor: expected text string")
    }
    if (minor > rest.length) {
        throw Error("cbor: unexpected EOF")
    }
    return [new TextDecoder().decode(rest.subarray(0, minor)), rest.subarray(minor)]
}

export function readArray(b: Uint8Array): [string[], Uint8Array] {
    const [major, minor, r] = readTypeAndArgument(b)
    if (major !== 4) {
        throw Error("cbor: expected array")
    }
    let rest = r
    const args = []
    for (let i = 0; i < minor; i++) {
        let arg
        [arg, rest] = readTextString(rest)
        args.push(arg)
    }
    return [args, rest]
}

export function encodeUint(n: number): Uint8Array {
    if (n <= 23) {
        return new Uint8Array([n])
    }
    if (n <= 0xff) {
        return new Uint8Array([24, n])
    }
    if (n <= 0xffff) {
        return new Uint8Array([25, n >> 8, n & 0xff])
    }
    throw Error("cbor: unsigned integer too large")
}

export function encodeByteString(b: Uint8Array): Uint8Array {
    if (b.length <= 23) {
        return new Uint8Array([2 << 5 | b.length, ...b])
    }
    if (b.length <= 0xff) {
        return new Uint8Array([2 << 5 | 24, b.length, ...b])
    }
    if (b.length <= 0xffff) {
        return new Uint8Array([2 << 5 | 25, b.length >> 8, b.length & 0xff, ...b])
    }
    throw Error("cbor: byte string too long")
}

export function encodeTextString(s: string): Uint8Array {
    const b = new TextEncoder().encode(s)
    if (b.length <= 23) {
        return new Uint8Array([3 << 5 | b.length, ...b])
    }
    if (b.length <= 0xff) {
        return new Uint8Array([3 << 5 | 24, b.length, ...b])
    }
    if (b.length <= 0xffff) {
        return new Uint8Array([3 << 5 | 25, b.length >> 8, b.length & 0xff, ...b])
    }
    throw Error("cbor: text string too long")
}

export function encodeArray(args: string[]): Uint8Array {
    const body = args.flatMap(x => [...encodeTextString(x)])
    if (args.length <= 23) {
        return new Uint8Array([4 << 5 | args.length, ...body])
    }
    if (args.length <= 0xff) {
        return new Uint8Array([4 << 5 | 24, args.length, ...body])
    }
    if (args.length <= 0xffff) {
        return new Uint8Array([4 << 5 | 25, args.length >> 8, args.length & 0xff, ...body])
    }
    throw Error("cbor: array too long")
}

File: fido2prf/internal/ctap2cbor/cbor.go
// Package ctap2cbor implements a tiny subset of CTAP2's subset of CBOR,
// in order to encode and decode WebAuthn identities.
//
// Only major types 0 (unsigned integer), 2 (byte strings), 3 (text strings),
// and 4 (arrays, only containing text strings) are supported. Arguments are
// limited to 16-bit values.
//
// See https://www.imperialviolet.org/tourofwebauthn/tourofwebauthn.html#cbor.
package ctap2cbor

import "encoding/binary"

func appendTypeAndArgument(buf []byte, major uint8, arg uint16) []byte {
	if arg <= 23 {
		return append(buf, (major<<5)|uint8(arg))
	}
	if arg <= 0xff {
		return append(buf, (major<<5)|24, uint8(arg))
	}
	return append(buf, (major<<5)|25, uint8(arg>>8), uint8(arg))
}

func AppendUint(buf []byte, arg uint16) []byte {
	return appendTypeAndArgument(buf, 0, arg)
}

func AppendBytes(buf []byte, arg []byte) []byte {
	if len(arg) > 0xffff {
		panic("ctap2cbor: byte string too long")
	}
	buf = appendTypeAndArgument(buf, 2, uint16(len(arg)))
	return append(buf, arg...)
}

func AppendString(buf []byte, arg string) []byte {
	if len(arg) > 0xffff {
		panic("ctap2cbor: string too long")
	}
	buf = appendTypeAndArgument(buf, 3, uint16(len(arg)))
	return append(buf, arg...)
}

func AppendArray(buf []byte, arg ...string) []byte {
	if len(arg) > 0xffff {
		panic("ctap2cbor: array too long")
	}
	buf = appendTypeAndArgument(buf, 4, uint16(len(arg)))
	for _, s := range arg {
		buf = AppendString(buf, s)
	}
	return buf
}

type String []byte

func (s String) Empty() bool {
	return len(s) == 0
}

func (s *String) readTypeAndArgument() (major uint8, arg uint16, ok bool) {
	if len(*s) < 1 {
		return
	}
	major = (*s)[0] >> 5
	minor := (*s)[0] & 0x1f
	switch {
	case minor <= 23:
		arg = uint16(minor)
		*s = (*s)[1:]
	case minor == 24:
		if len(*s) < 2 {
			return
		}
		arg = uint16((*s)[1])
		*s = (*s)[2:]
	case minor == 25:
		if len(*s) < 3 {
			return
		}
		arg = binary.BigEndian.Uint16((*s)[1:])
		*s = (*s)[3:]
	default:
		return
	}
	ok = true
	return
}

func (s *String) ReadUint(out *uint16) bool {
	major, arg, ok := s.readTypeAndArgument()
	if !ok || major != 0 {
		return false
	}
	*out = arg
	return true
}

func (s *String) ReadBytes(out *[]byte) bool {
	major, arg, ok := s.readTypeAndArgument()
	if !ok || major != 2 {
		return false
	}
	if len(*s) < int(arg) {
		return false
	}
	*out = (*s)[:arg]
	*s = (*s)[arg:]
	return true
}

func (s *String) ReadString(out *string) bool {
	major, arg, ok := s.readTypeAndArgument()
	if !ok || major != 3 {
		return false
	}
	if len(*s) < int(arg) {
		return false
	}
	*out = string((*s)[:arg])
	*s = (*s)[arg:]
	return true
}

func (s *String) ReadArray(out *[]string) bool {
	major, arg, ok := s.readTypeAndArgument()
	if !ok || major != 4 {
		return false
	}
	for i := range arg {
		*out = append(*out, "")
		if !s.ReadString(&(*out)[i]) {
			return false
		}
	}
	return true
}
